gdal/ogr/ogrsf_frmts/cad/gdalcaddataset.cpp

```diff
@@ -0,0 +1,463 @@
+}
+
+GDALCADDataset::~GDALCADDataset()
+    {
+    if (poRasterDS != NULL)
+    {
+        GDALClose( poRasterDS );
+        poRasterDS = NULL;
+    }
+    
+    for( int i = 0; i < nLayers; i++ )
+        delete papoLayers[i];
+    CPLFree( papoLayers );
+    if(poCADFile)
+        delete( poCADFile );
+}
+    
+void GDALCADDataset::FillTransform(CADImage* pImage, double dfUnits)
+{
+            unsigned char nResUnits = pImage->getResolutionUnits();
+            double dfMultiply(1);
+            
+            switch(nResUnits)// 0 == none, 2 == centimeters, 5 == inches;
+            {
+                case 2:
+            dfMultiply = 100 / dfUnits; // meters to linear units
+            break;
+                case 5: 
+            dfMultiply = 0.0254 / dfUnits;   
+            break;
+                case 0:                
+                default:
+                    dfMultiply = 1;
+            }
+            
+    CADVector oSizePt = pImage->getImageSizeInPx();
+    CADVector oInsPt = pImage->getVertInsertionPoint();
+    CADVector osSizeUnitsPt = pImage->getPixelSizeInACADUnits();
+    adfGeoTransform[0] = oInsPt.getX();
+    adfGeoTransform[3] = oInsPt.getY() + oSizePt.getY() * osSizeUnitsPt.getX() * dfMultiply;
+    adfGeoTransform[2] = 0;
+    adfGeoTransform[4] = 0;
+
+            adfGeoTransform[1] = osSizeUnitsPt.getX() * dfMultiply;
+            adfGeoTransform[5] = -osSizeUnitsPt.getY() * dfMultiply;
+}
+
+int GDALCADDataset::Open( GDALOpenInfo* poOpenInfo, CADFileIO* pFileIO,
+                                    long nSubRasterLayer, long nSubRasterFID )
+{
+    size_t i, j;
+    int nRasters = 1;
+
+    osCADFilename = pFileIO->GetFilePath();
+    SetDescription( poOpenInfo->pszFilename );
+    
+    int nMode = atoi(CSLFetchNameValueDef( poOpenInfo->papszOpenOptions, "MODE", 
+                                                                           "2"));
+    
+    enum CADFile::OpenOptions openOpts = CADFile::READ_FAST;
+    if(nMode == 1)
+    {
+        openOpts = CADFile::READ_ALL;
+    }
+    else if(nMode == 3)
+    {
+        openOpts = CADFile::READ_FASTEST;
+    }
+    
+    poCADFile = OpenCADFile( pFileIO, openOpts );
+
+    if ( GetLastErrorCode() == CADErrorCodes::UNSUPPORTED_VERSION )
+    {
+        CPLError( CE_Failure, CPLE_NotSupported,
+                  "libopencad %s does not support this version of CAD file.\n"
+                  "Supported formats are:\n%s", GetVersionString(), GetCADFormats() );
+        return( FALSE );
+    }
+    
+    if ( GetLastErrorCode() != CADErrorCodes::SUCCESS )
+    {
+        CPLError( CE_Failure, CPLE_NotSupported,
+                  "libopencad %s does not support this version of CAD file.\nSupported formats: %s", 
+                  GetVersionString(), GetCADFormats() );
+        return( FALSE );
+    }
+    
+
+    OGRSpatialReference *poSpatialRef = GetSpatialReference( );
+
+    if( nSubRasterLayer != -1 && nSubRasterFID != -1 )
+    {
+        nRasters = 2;
```

Comment:
Yes, this is some kind of magic number. If we have we only one raster layer in DWG or special subdataset the nRasters == 2. 

There are 2 cases: 
1. CADDataset open DWG file with only one raster layer. So because this:
https://github.com/sandyre/gdal_libopencad/blob/trunk/gdal/ogr/ogrsf_frmts/cad/gdalcaddataset.cpp#L196. nRasters == 2
2. CADDataset open subdataset and receive not -1 for nSubRasterLayer and nSubRasterFID

To return raster in both cases nRaster mast be equal 2.
We can add more complicated logic to change nRaster to 1 and rename it to nRasterCount, but should it?